### PR TITLE
[FLINK-22678] Configurations and user APIs for ChangelogStateBackend

### PR DESCRIPTION
**Pull request by Zakelly**
## What is the purpose of the change

This change provides configurations as well as APIs to configure ChangelogStateBackend. Base on the discussion before (https://lists.apache.org/thread.html/ra178ce29088b1da362d98a5a6d8c7be48051caf1637ee24261738217%40%3Cdev.flink.apache.org%3E) and the design doc (https://docs.google.com/document/d/13AaCf5fczYTDHZ4G1mgYL685FqbnoEhgo0cdwuJlZmw/edit?usp=sharing, see Option 3), we make a refined version of loading ChangelogStateBackend. Please check the brief change log below.

## Brief change log

- Define consistent override and combination policy (flag + state backend) in different config levels.
- Differentiate explicitly the meaning of "enable flag" = true/false/unset.
- Hide ChangelogStateBackend from users.

## Verifying this change

This change added some test cases in ```ChangelogStateBackendLoadingTest```. Including tests for configuration overriding and stream graph validation.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes, tests in ```flink-statebackend-changelog``` now depends on ```flink-streaming-java_2.11```.
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes, add ```enableChangelogStateBackend``` and ```isChangelogStateBackendEnabled``` in ```StreamExecutionEnvironment```
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no, but ChangelogStateBackend itself may affect the per-record code paths.
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no
 
## Documentation
  - Does this pull request introduce a new feature? yes.
  - If yes, how is the feature documented? Docs and JavaDocs.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -593,7 +598,20 @@ public CheckpointingMode getCheckpointingMode() {
      */
     @PublicEvolving
     public StreamExecutionEnvironment setStateBackend(StateBackend backend) {
-        this.defaultStateBackend = Preconditions.checkNotNull(backend);
+        Preconditions.checkNotNull(backend);
+        if (changelogStateBackendEnabled.equals(TernaryBoolean.TRUE)) {
+            try {
+                this.defaultStateBackend =
```
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:604]: I don't think we should load the changelog state backend here, we could still keep the original internal state backend and load the real one within `StreamGraphGenerator` or `StreamTask`.
Moreover, if so, there is no need to introduce `StateBackendLoader#loadChangelogStateBackendIfNeeded` and `StateBackendLoader#unwrapFromDelegatingStateBackend`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java [hunk 2]
```diff
@@ -327,6 +331,7 @@ private static StateBackend loadFromApplicationOrConfigOrDefaultInternal(
      */
     public static StateBackend fromApplicationOrConfigOrDefault(
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:332]: We should add logs to tell that whether changelog state backend is enabled.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java [hunk 3]
```diff
@@ -336,8 +341,18 @@ public static StateBackend fromApplicationOrConfigOrDefault(
                 loadFromApplicationOrConfigOrDefaultInternal(
                         fromApplication, config, classLoader, logger);
 
-        if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)
-                && !(fromApplication instanceof DelegatingStateBackend)) {
+        if (isChangeLogStateBackendEnableFromApplication == null) {
+            isChangeLogStateBackendEnableFromApplication = TernaryBoolean.UNDEFINED;
+        }
+        Optional<Boolean> enableConfig =
+                config.getOptional(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG);
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:348]: Why not use `Boolean enabledFromConfig = config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG);` directly?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java [hunk 4]
```diff
@@ -327,6 +331,7 @@ private static StateBackend loadFromApplicationOrConfigOrDefaultInternal(
      */
     public static StateBackend fromApplicationOrConfigOrDefault(
             @Nullable StateBackend fromApplication,
+            @Nullable TernaryBoolean isChangeLogStateBackendEnableFromApplication,
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:334]: If `isChangeLogStateBackendEnableFromApplication` is `null`, it actually has the meaning as `unset` (mainly on stream task side). However, `TernaryBoolean` itself is designed to be able to present the `unset` meaning. I think here existed mixed semantics and I suggest to remove the possibility of `Nullable`.

File: flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java [hunk 5]
```diff
@@ -183,16 +218,83 @@ public void testLoadingRocksDBStateBackend() throws Exception {
                 false);
     }
 
-    private Configuration config(String stateBackend) {
-        final Configuration config = config();
+    @Test
+    public void testConfigureEnvByConfiguration() throws Exception {
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java:222]: If we decide to remove loading state backend in streaming execution environment, the methods of `testConfigureEnvByConfiguration` and `testEnableChangelogStateBackendInStreamExecutionEnvironment` are useless.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 6]
```diff
@@ -607,6 +625,65 @@ public StateBackend getStateBackend() {
...
+     *     the change log.
+     * @return This StreamExecutionEnvironment itself, to allow chaining of function calls.
+     * @see #isChangelogStateBackendEnabled()
+     */
+    @PublicEvolving
+    public StreamExecutionEnvironment enableChangelogStateBackend(boolean enabled) {
+        if (enabled) {
+            try {
+                this.defaultStateBackend =
+                        StateBackendLoader.loadChangelogStateBackendIfNeeded(
```
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:661]: I don't think we should load state backend once we enable changelog state-backend as client-side might have different classloader with cluster-runtime-side. Just leave the original internal state backend and load it once running on cluster.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/ExecutorUtils.java [hunk 7]
```diff
@@ -73,6 +75,7 @@ public static void setBatchProperties(StreamGraph streamGraph, TableConfig table
         // scheduler)
         streamGraph.setJobType(JobType.BATCH);
         streamGraph.setStateBackend(null);
+        streamGraph.setChangeLogStateBackendEnabled(TernaryBoolean.UNDEFINED);
```
Comment (Myasuka) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/ExecutorUtils.java:78]: If no state-backend is needed for batch jobs, should we set set enable changlog as `TernaryBoolean.FALSE`?

**Review by Myasuka — COMMENTED** [review]
@Zakelly , thanks for your contributions!

Please take a look at my comments.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java [hunk 8]
```diff
@@ -214,6 +215,8 @@ StreamConfig getConfig(
         config.setOperatorName(operatorID.toHexString());
         config.setOperatorID(operatorID);
         config.setStateBackend(stateBackend);
+        // This means leave this stateBackend unwrapped.
```
[the remark this answers is not in the document]
Comment (Zakelly) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java:218]: fixed

**Review by Zakelly — COMMENTED** [review]
@Myasuka, thanks for the detailed review.

File: flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java [hunk 9]
```diff
@@ -102,13 +102,15 @@
 
     /** Whether to enable state change log. */
     @Documentation.Section(value = Documentation.Sections.COMMON_STATE_BACKENDS)
-    @Documentation.ExcludeFromDocumentation("Hidden for now")
     public static final ConfigOption<Boolean> ENABLE_STATE_CHANGE_LOG =
             ConfigOptions.key("state.backend.changelog.enabled")
                     .booleanType()
-                    .defaultValue(false)
+                    .noDefaultValue()
```
Comment (Myasuka) [inline flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:108]: As discussed offline, this configuration should have a default value of `false`. We might have different kinds of implementation but we should tell user explicitly that this option is disabled by default.

**Review by Myasuka — COMMENTED** [review]
Thanks for update! And please take a look at my comments.

File: flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java (hunk 9 above)
Comment (Zakelly) [reply flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java]: Yes, the default value here is false, but the return value of ```StreamExecutionEnvironment::isChangelogStateBackendEnabled``` is a ```TernaryBoolean```, which may bring some confusion. The ```TernaryBoolean``` is essential cause table planner uses the value to create StreamGraphGenerator (see [here](https://github.com/apache/flink/pull/16153/commits/2c78381d58d1e7e3ee3edd06afb03e7e6eb1f177#diff-c2ed4f07637221036620a06234633cbed96eedf4f29297fd3d766b3e1e751cbaR48)). @rkhachatryan and @curcur would you please to take a look?

**Comment by Zakelly** [comment]
@Myasuka , thanks for the review! Would you mind taking another look? Thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java [hunk 10]
```diff
@@ -175,21 +174,16 @@ public static CheckpointStorage load(
...
         }
 
-        // Legacy state backends always take precedence for backwards compatibility.
-        StateBackend rootStateBackend =
-                (configuredStateBackend instanceof DelegatingStateBackend)
-                        ? ((DelegatingStateBackend) configuredStateBackend)
-                                .getDelegatedStateBackend()
-                        : configuredStateBackend;
-
-        if (rootStateBackend instanceof CheckpointStorage) {
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:185]: Could you please explain the removal? 
IIUC, changelog state backend can still be passed here (even if hidden from user) when called from `StreamTask#createCheckpointStorage`

cc: @curcur

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java [hunk 11]
```diff
@@ -384,6 +358,22 @@ public static boolean stateBackendFromApplicationOrConfigOrDefaultUseManagedMemo
...
+    /**
+     * Try to unwrap a DelegatingStateBackend and get the inner delegated StateBackend. If the
+     * provided StateBackend is not a delegated StateBackend, just return itself.
+     *
+     * @param backend the provided StateBackend that may delegate another StateBackend.
+     * @return the root StateBackend W/O delegation.
+     */
+    @VisibleForTesting
+    public static StateBackend unwrapFromDelegatingStateBackend(StateBackend backend) {
+        if (backend != null && backend instanceof DelegatingStateBackend) {
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:370]: nit: Is this method used **only** for tests? If so, maybe move it to `test` source directory?
nit: I think `null` check is unnecessary here.

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java [hunk 12]
```diff
@@ -58,7 +58,14 @@
...
 
-    public ChangelogStateBackend(StateBackend stateBackend) {
+    /**
+     * Delegate a state backend by a ChangelogStateBackend.
+     *
+     * <p>As FLINK-22678 mentioned, we currently hide this constructor from user.
+     *
+     * @param stateBackend the delegated state backend.
+     */
+    ChangelogStateBackend(StateBackend stateBackend) {
```
Comment (rkhachatryan) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:68]: Maybe also mark the class as `@Internal` to make the intention clear to the user?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 13]
```diff
@@ -607,6 +612,53 @@ public StateBackend getStateBackend() {
...
+     * log. Configs for change log enabling will override in different config levels
+     * (job/local/cluster).
+     *
+     * @param enabled true if enable the change log for state backend explicitly, otherwise disable
+     *     the change log.
+     * @return This StreamExecutionEnvironment itself, to allow chaining of function calls.
+     * @see #isChangelogStateBackendEnabled()
+     */
+    @PublicEvolving
+    public StreamExecutionEnvironment enableChangelogStateBackend(boolean enabled) {
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:644]: very good description :+1: :slightly_smiling_face: 

Should we also mention that it only applies to keyed state currently? (so non-keyed operator state and channel state are checkpointed as usual)

**Review by rkhachatryan — COMMENTED** [review]
Thanks for the PR @Zakelly.
I left some comments, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java (hunk 10 above)
Comment (Zakelly) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java]: You are right. This should not be removed. I made a mistake.
I revert this change. Thanks a lot for pointing this out.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java (hunk 11 above)
Comment (Zakelly) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java]: Sure, moved to ```ChangelogStateBackendLoadingTest```.

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java (hunk 12 above)
Comment (Zakelly) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:70]: Sure, I apply ```@Internal``` to class.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 13 above)
Comment (Zakelly) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:646]: Yes, I added some description. 🙂

**Comment by Zakelly** [comment]
@rkhachatryan thanks for the review! I've addressed your comments. Thanks!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 13 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:646]: Thanks! Could you please do the same for the scala version? 
Sorry for clarifying in the beginning.

Comment (Zakelly) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:646]: Sure thing, at [1ef3786](https://github.com/apache/flink/pull/16153/commits/1ef3786ca933d1bf675706d2f4f204314f3b540b).
And as for the python datastream API, I left it unimplemented. I'll create another PR for it.

**Review by rkhachatryan — APPROVED** [review]
Thanks for updating the PR @Zakelly!
LGTM.

File: flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java [hunk 14]
```diff
@@ -102,13 +102,17 @@
...
             ConfigOptions.key("state.backend.changelog.enabled")
                     .booleanType()
                     .defaultValue(false)
                     .withDescription(
-                            "Whether to enable state backend to write state changes to StateChangelog.");
+                            "Whether to enable state backend to write state changes to StateChangelog. "
+                                    + "If this config is not set explicitly, it means no preference "
+                                    + "for enabling the change log, and the value in lower config "
+                                    + "level will take effect. The default value 'false' here means "
+                                    + "if no value set (job or cluster), the change log will not be"
```
Comment (zhangyy91) [inline flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:114]: @Zakelly Hi, I reviewed the pull request and, in general LGTM.
I found a minor error in the doc string. There should be a space at the end of the line. Or the generated doc will be "the change log will not beenabled".

Comment (Zakelly) [reply flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java]: Yes, that's right. Thanks a lot for pointing this out! I have fixed this at [080b0ab](https://github.com/apache/flink/pull/16153/commits/080b0ab09afc18ee5e24bd8b536380273cae0d4c)

Thanks again!

**Pull request merged**